### PR TITLE
Move serviceworker to workbox and fix SSE interference

### LINKED ISSUE #7372: reference error in PWA serviceworker
`ReferenceError: caches is not defined` - serviceworker.js, line 60.

Appeared in #4730 PR, reproducible on https://try.gitea.io if you load Gitea **in Firefox.** Firefox doesn't have the `caches` variable.

### LINKED ISSUE #11092: Automate serviceworker caching manifest generation
The script at https://github.com/go-gitea/gitea/blob/master/templates/pwa/serviceworker_js.tmpl should be replaced with something that automatically generates a appropriate service worker caching manifest based on the present files in `public` at compile time and then use that data together with  `StaticUrlPrefix` at runtime to serve the actual URLs.

Normally, one would use a webpack plugin for this, but webpack can not know about `StaticUrlPrefix`, and also does not know about `vendor` files so it's not directly suitable, but could maybe be integrated in the compile-time part of the process.

### PR DESCRIPTION
**Pull request by silverwind**
Instead of statically hardcoding every frontend asset, this uses a type-based approach to cache all js,css and manifest.json requests.

This also fixes the issue that the service worker was interfering with EventSource because it was unconditionally handling all requests which this new implementation doesn't.

Fixes: https://github.com/go-gitea/gitea/issues/11092
Fixes: https://github.com/go-gitea/gitea/issues/7372

### CODE REVIEW

**Comment by CirnoT** [comment]
Confirmed that this solves issue with EventSource

**Review by CirnoT — APPROVED** [review]
Obviously can not test every edge case but fixes my issues with SW and populates cache properly as expected.

**Review by lafriks — APPROVED** [review]

**Comment by silverwind** [comment]
Should we remove `templates.JSRenderer`? I think it'll be unused after this PR.

**Comment by guillep2k** [comment]
> 
> 
> Should we remove `templates.JSRenderer`? I think it'll be unused after this PR and I don't see us rendering JS templates in the near future 😉.

In a separate PR, please.

**Comment by silverwind** [comment]
@guillep2k are you sure? I think it's fine to remove here as it was only used for the serviceworker rendering and added in the [original commit](https://github.com/go-gitea/gitea/commit/294904321cb6de535237a6a156d5c4ec462bc117) so I think it only makes sense to remove it now.

**Comment by silverwind** [comment]
Removed it. @guillep2k hope it's okay for you 😉

**Comment by silverwind** [comment]
One more feature added: Cache will now invalidate if `AppVer` (e.g. `1.13.0+dev-43-g8636e470d`) changes. It keeps the version in `localStorage` for this purpose.

`ui.USE_SERVICE_WORKER` is still required for development as `AppVer` does not change between frontend rebuilds.

**Comment by techknowlogick** [comment]
ping LG-TM

**Comment by techknowlogick** [comment]
@silverwind please send backport to release/v1.12 :)

**Comment by silverwind** [comment]
Backport done.

I'll follow up with another PR to this, there are some minor fixups to be done, but nothing crucial 😉

File: web_src/js/features/serviceworker.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  }
+}
+
+export default async function initServiceWorker() {
+  if (!('serviceWorker' in navigator)) return;
+
+  if (UseServiceWorker) {
+    await checkCacheValidity();
+    try {
+      await navigator.serviceWorker.register(`${AppSubUrl}/serviceworker.js`);
```
Comment (lunny) [inline web_src/js/features/serviceworker.js:33]: Isn't the router has been removed from macaron? Maybe you mean `${StaticUrlPrefix}/js/serviceworker.js`?

**Comment by lunny** [comment]
Please notice all js files may need to be cached by CDN.

**Pull request merged**